src/v/raft/seq_stm.h

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright 2020 Vectorized, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "kafka/protocol/errors.h"
+#include "model/fundamental.h"
+#include "model/record.h"
+#include "raft/consensus.h"
+#include "raft/errc.h"
+#include "raft/logger.h"
+#include "raft/state_machine.h"
+#include "raft/types.h"
+#include "storage/snapshot.h"
+#include "utils/expiring_promise.h"
+#include "utils/mutex.h"
+
+#include <absl/container/flat_hash_map.h>
+
+namespace raft {
+
+struct seq_snapshot_header {
+    static constexpr const int8_t supported_version = 0;
+
+    int8_t version;
+    int32_t snapshot_size;
+
+    static constexpr const size_t ondisk_size = sizeof(version)
+                                                + sizeof(snapshot_size);
+};
+
+class seq_stm final
+  : public raft::state_machine
+  , public storage::stm_manager {
+public:
+    using replication_result
+      = checked<raft::replicate_result, kafka::error_code>;
+    explicit seq_stm(ss::logger&, raft::consensus*, config::configuration&);
+
+    ss::future<> start() final;
+
+    ss::future<> ensure_snapshot_exists(model::offset) final;
+    ss::future<> make_snapshot() final;
+    ss::future<> catchup();
+
+    ss::future<checked<raft::replicate_result, kafka::error_code>> replicate(
+      raft::batch_identity,
+      model::record_batch_reader&&,
+      raft::replicate_options);
+
+private:
+    struct seq_entry {
```

Comment:
You always specify all the fields, so why not require the constructor to specify all the fields so that in a future refactoring you don't leave anything out by mistake.

Reply:
The problem with this is that I use seq_entry also to serialize a snapshot to disk using `adl.h` and it depends on structs having brace constructible to detect arity. Once I add a constructor it removes brace constructible pattern and breaks serialization. I could add another struct to act as a medium but I don't feel it's worth the hassle since seq_entry is private to seq_stm so it's scope is limited and a risk of typo is low

 What do you think?

Reply:
sounds good.